// babel-plugin-transform-async-generator-functions/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = _default;

var _babelHelperRemapAsyncToGenerator = _interopRequireDefault(require("babel-helper-remap-async-to-generator"));

var _babelPluginSyntaxAsyncGenerators = _interopRequireDefault(require("babel-plugin-syntax-async-generators"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default(_ref) {
  var t = _ref.types;
  var yieldStarVisitor = {
    Function: function Function(path) {
      path.skip();
    },
    YieldExpression: function YieldExpression(_ref2, state) {
      var node = _ref2.node;
      if (!node.delegate) return;
      var callee = state.addHelper("asyncGeneratorDelegate");
      node.argument = t.callExpression(callee, [t.callExpression(state.addHelper("asyncIterator"), [node.argument]), t.memberExpression(state.addHelper("asyncGenerator"), t.identifier("await"))]);
    }
  };
  return {
    inherits: _babelPluginSyntaxAsyncGenerators.default,
    visitor: {
      Function: function Function(path, state) {
        if (!path.node.async || !path.node.generator) return;
        path.traverse(yieldStarVisitor, state);
        (0, _babelHelperRemapAsyncToGenerator.default)(path, state.file, {
          wrapAsync: t.memberExpression(state.addHelper("asyncGenerator"), t.identifier("wrap")),
          wrapAwait: t.memberExpression(state.addHelper("asyncGenerator"), t.identifier("await"))
        });
      }
    }
  };
}
